src/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,208 @@
+                              << std::endl;
+                }
+                break;
+            }
+
+            numDecoded++;
+
+            m_log_events.push_back(result.value());
+        }
+
+        return numDecoded;
+    }
+
+    bool decode(const emscripten::val &results, size_t startIdx, size_t endIdx)
+    {
+        for (size_t log_event_idx = startIdx; log_event_idx < endIdx; log_event_idx++)
+        {
+            auto const &log_event{m_log_events[log_event_idx]};
+            std::string message;
+
+            {
+                auto constant_handler = [&](std::string const &value, size_t begin_pos, size_t length)
+                {
+                    message.append(value, begin_pos, length);
+                };
+
+                auto encoded_int_handler = [&](clp::ir::four_byte_encoded_variable_t value)
+                { message.append(clp::ffi::decode_integer_var(value)); };
+
+                auto encoded_float_handler = [&](clp::ir::four_byte_encoded_variable_t encoded_float)
+                {
+                    message.append(clp::ffi::decode_float_var(encoded_float));
+                };
+
+                auto dict_var_handler = [&](std::string const &dict_var)
+                { message.append(dict_var); };
+
+                try
+                {
+                    clp::ffi::ir_stream::generic_decode_message<true>(
+                        log_event.get_logtype(),
+                        log_event.get_encoded_vars(),
+                        log_event.get_dict_vars(),
+                        constant_handler,
+                        encoded_int_handler,
+                        encoded_float_handler,
+                        dict_var_handler);
+                }
+                catch (clp::ffi::ir_stream::DecodingException const &e)
+                {
+                    std::cerr << "Failed to decode. Error Code: " << static_cast<uint32_t>(clp::ffi::ir_stream::IRErrorCode_Decode_Error)
+                              << std::endl;
+
+                    break;
+                }
+            }
+
+            LOG_VERBOSITY verbosity_level = LOG_VERBOSITY::NONE;
+            for (size_t verbosity_idx = 1; verbosity_idx < sizeof(VERBOSITIES) / sizeof(VERBOSITIES[0]); verbosity_idx++)
+            {
+                if (log_event.get_logtype().substr(1).starts_with(VERBOSITIES[verbosity_idx]))
+                {
+                    verbosity_level = static_cast<LOG_VERBOSITY>(verbosity_idx);
+                    break;
+                }
+            }
+
+            m_ts_pattern.insert_formatted_timestamp(
+                log_event.get_timestamp() + m_timezone_offset_ms,
+                message);
+
+            insert_row(results.as_handle(), message.c_str(), log_event.get_timestamp(),
+                       static_cast<int>(verbosity_level));
+        }
+
+        return true;
+    }
+
+private:
+    bool decodePreamble()
```

Comment:
Instead of having this decode_preamble as a separate function, can we make it RAII? This means we will deserialize the preamble in the constructor, and ensure all member variables are initialized inside the constructor.
By thinking a little further, is it possible to use a factory function instead? By doing this we don't have to use unique ptr or std::optional for deserializer. But that also means js will need to call a method to create an object, instead of creating through a constructor. Feel free to chat offline.

Reply:
RAII makes sense. A factory method is fine but will also require changing the interfaces in our existing JS code which is less ideal. I'll look more into those.

Reply:
I think from code design perspective a factory function might make more sense as it avoid using unique ptr